// cz-conventional-changelog/index.js

"format cjs";

var wrap = require('word-wrap');

// This can be any kind of SystemJS compatible module.
// We use Commonjs here, but ES6 or AMD would do just
// fine.
module.exports = {

  // When a user runs `git cz`, prompter will
  // be executed. We pass you cz, which currently
  // is just an instance of inquirer.js. Using
  // this you can ask questions and get answers.
  //
  // The commit callback should be executed when
  // you're ready to send back a commit template
  // to git.
  //
  // By default, we'll de-indent your commit
  // template and will keep empty lines.
  prompter: function(cz, commit) {
    console.log('\nLine 1 will be cropped at 100 characters. All other lines will be wrapped after 100 characters.\n');

    // Let's ask some questions of the user
    // so that we can populate our commit
    // template.
    //
    // See inquirer.js docs for specifics.
    // You can also opt to use another input
    // collection library if you prefer.
    cz.prompt([
      {
        type: 'list',
        name: 'type',
        message: '选择您正在提交的内容:',
        choices: [
        {
          name: 'Feature: 新功能添加',
          value: 'feat'
        }, {
          name: 'Fix bug: 修补bug',
          value: 'fix'
        }, {
          name: 'Documentation: 文档添加及修改',
          value: 'docs'
        }, {
          name: 'Style: 不影响功能的代码变化（空格，格式化，丢失分号等）',
          value: 'style'
        }, {
          name: 'Refactor:    代码重构',
          value: 'refactor'
        }, {
          name: 'Performance improves:    提高性能的代码更改',
          value: 'perf'
        }, {
          name: 'Add test:    添加测试代码，或功能测试代码添加',
          value: 'test'
        }, {
          name: 'Auxiliary tools:    改变构建过程或辅助工具和库，如文档生成',
          value: 'chore'
        }]
      }, {
        type: 'input',
        name: 'scope',
        message: '用来说明本次Commit影响的范围，即简要说明修改会涉及的部分。:\n'
      }, {
        type: 'input',
        name: 'subject',
        message: '用来简要描述本次改动，概述就好了，因为后面还会在Body里给出具体信息:\n'
      }, {
        type: 'input',
        name: 'body',
        message: '对上面subject里内容的展开，在此做更加详尽的描述，内容里应该包含修改动机和修改前后的对比:\n'
      }, {
        type: 'input',
        name: 'footer',
        message: '不兼容变更和Issue关闭的信息:\n,如:  Closes #56, #25'
      }
    ]).then(function(answers) {

      var maxLineWidth = 100;

      var wrapOptions = {
        trim: true,
        newline: '\n',
        indent:'',
        width: maxLineWidth
      };

      // parentheses are only needed when a scope is present
      var scope = answers.scope.trim();
      scope = scope ? '(' + answers.scope.trim() + ')' : '';

      // Hard limit this line
      var head = (answers.type + scope + ': ' + answers.subject.trim()).slice(0, maxLineWidth);

      // Wrap these lines at 100 characters
      var body = wrap(answers.body, wrapOptions);
      var footer = wrap(answers.footer, wrapOptions);

      commit(head + '\n\n' + body + '\n\n' + footer);
    });
  }
}
